// twilio-mcs-client/browser/media.js
"use strict";

var _regenerator = require("babel-runtime/regenerator");

var _regenerator2 = _interopRequireDefault(_regenerator);

var _promise = require("babel-runtime/core-js/promise");

var _promise2 = _interopRequireDefault(_promise);

var _asyncToGenerator2 = require("babel-runtime/helpers/asyncToGenerator");

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _classCallCheck2 = require("babel-runtime/helpers/classCallCheck");

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require("babel-runtime/helpers/createClass");

var _createClass3 = _interopRequireDefault(_createClass2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

Object.defineProperty(exports, "__esModule", { value: true });
exports.Media = void 0;
/**
 * @classdesc A Media represents a metadata information for the media upload
 * @property {String} sid - The server-assigned unique identifier for Media
 * @property {String} serviceSid - Service instance id which Media belongs/uploaded to
 * @property {Date} dateCreated - When the Media was created
 * @property {Date} dateUpdated - When the Media was updated
 * @property {Number} size - Size of media, bytes
 * @property {String} contentType - content type of media
 * @property {String} fileName - file name, if present, null otherwise
 * @property {MediaCategory} category - attachment category
 */

var Media = function () {
    function Media(config, network, data) {
        (0, _classCallCheck3.default)(this, Media);

        this.config = config;
        this.network = network;
        this._update(data);
    }

    (0, _createClass3.default)(Media, [{
        key: "getContentUrl",

        /**
         * Returns direct content URL to uploaded binary. This URL will expire after some time.
         * This function gets a new URL every time, preventing it from expiring but putting additional load on backend.
         * See getCachedContentUrl() for a function that reduces the amount of network requests.
         *
         * It is reasonable to build your own refresh logic upon these two functions: as soon as URL returned
         * by getCachedContentUrl() returns 40x status you should call getContentUrl() to refresh it.
         *
         * @returns {Promise<string>}
         */
        value: function () {
            var _ref = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee() {
                var response;
                return _regenerator2.default.wrap(function _callee$(_context) {
                    while (1) {
                        switch (_context.prev = _context.next) {
                            case 0:
                                _context.next = 2;
                                return this.network.get(this.config.baseUrl + "/" + this.sid);

                            case 2:
                                response = _context.sent;

                                this._update(response.body);
                                return _context.abrupt("return", _promise2.default.resolve(this.state.contentDirectUrl));

                            case 5:
                            case "end":
                                return _context.stop();
                        }
                    }
                }, _callee, this);
            }));

            function getContentUrl() {
                return _ref.apply(this, arguments);
            }

            return getContentUrl;
        }()
        /**
         * Returns direct content URL to uploaded binary. This URL will expire after some time.
         * This function does not support getting a new URL however. Once set it will remain the same.
         * Use getContentUrl() to query a new one.
         *
         * It is reasonable to build your own refresh logic upon these two functions: as soon as URL returned
         * by getCachedContentUrl() returns 40x status you should call getContentUrl() to refresh it.
         *
         * @returns {Promise<string>}
         */

    }, {
        key: "getCachedContentUrl",
        value: function () {
            var _ref2 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee2() {
                return _regenerator2.default.wrap(function _callee2$(_context2) {
                    while (1) {
                        switch (_context2.prev = _context2.next) {
                            case 0:
                                if (!this.state.contentDirectUrl) {
                                    _context2.next = 2;
                                    break;
                                }

                                return _context2.abrupt("return", _promise2.default.resolve(this.state.contentDirectUrl));

                            case 2:
                                _context2.next = 4;
                                return this.getContentUrl();

                            case 4:
                                return _context2.abrupt("return", _context2.sent);

                            case 5:
                            case "end":
                                return _context2.stop();
                        }
                    }
                }, _callee2, this);
            }));

            function getCachedContentUrl() {
                return _ref2.apply(this, arguments);
            }

            return getCachedContentUrl;
        }()
    }, {
        key: "_update",
        value: function _update(data) {
            var _a, _b, _c;
            this.state = {
                sid: data.sid,
                serviceSid: data.service_sid,
                channelSid: data.channel_sid,
                messageSid: data.message_sid,
                dateCreated: data.date_created ? new Date(data.date_created) : null,
                dateUploadUpdated: data.date_upload_updated ? new Date(data.date_upload_updated) : null,
                dateUpdated: data.date_updated ? new Date(data.date_updated) : null,
                size: data.size,
                contentType: data.content_type,
                author: data.author,
                url: data.url,
                contentUrl: data.links.content,
                contentDirectUrl: data.links.content_direct_temporary,
                filename: (_a = data.filename) !== null && _a !== void 0 ? _a : null,
                category: (_b = data.category) !== null && _b !== void 0 ? _b : 'media',
                isMultipartUpstream: (_c = data.is_multipart_upstream) !== null && _c !== void 0 ? _c : false
            };
        }
    }, {
        key: "sid",
        get: function get() {
            return this.state.sid;
        }
    }, {
        key: "serviceSid",
        get: function get() {
            return this.state.serviceSid;
        }
    }, {
        key: "dateCreated",
        get: function get() {
            return this.state.dateCreated;
        }
    }, {
        key: "dateUpdated",
        get: function get() {
            return this.state.dateUpdated;
        }
    }, {
        key: "contentType",
        get: function get() {
            return this.state.contentType;
        }
    }, {
        key: "size",
        get: function get() {
            return this.state.size;
        }
        /** @deprecated Use filename instead */

    }, {
        key: "fileName",
        get: function get() {
            return this.state.filename;
        }
    }, {
        key: "filename",
        get: function get() {
            return this.state.filename;
        }
    }, {
        key: "category",
        get: function get() {
            return this.state.category;
        }
    }]);
    return Media;
}();

exports.Media = Media;
